[pantsd] Implement auto-shutdown after runs

blorente opened the pull request:
In the context of enabling pantsd in Travis in some capacity, we need to implement a mechanism to auto-shutdown Pantsd after runs, so that we can still have isolated integration tests.

blorente commented:
(Sorry for converting an issue into a PR, I was testing out [hub](https://hub.github.com/). I don't know how to unconvert it, so I figured we could carry on here.)

### Problem

Described in this issue. We need a way to tell pantsd to shut down after running once.

### Solution

- Add a bootstrap flag `--shutdown-pantsd-after-run`, which tells pants to shut down after the first use.
- Read that flag in `PantsServices`, and pass it down to `PailgunService`.
- Implement a callback in `PailgunService` that terminates the service if the flag is enabled. Pipe that callback to `PailgunServer`. Any additional logic (e.g. "stop pantsd after x invocations instead of just one") should go here.
- Make `PailgunServer` call that callback unconditionally after every finished pants run.

Work to do:

- [ ] Add integration tests.
- [ ] Make pantsd close more gracefully when a service is shut down, instead of with an exception.

Result

Users can now use the flag, which has the following semantics:
```bash
./pants --enable-pantsd --shutdown-pantsd-after-run list # Close pantsd after running list
./pants --enable-pantsd list && ./pants --enable-panstd --shutdown-pantsd-after-run list
     # The new flag is fingerprinted, so changing it will trigger a daemon restart at the beginning of the second invocation, so the second invocation will have the same semantics as the first case.
./pants --enable-pantsd list && ./pants --shutdown-panstd-after-run list 
    # Disabling pantsd at the beginning of a run exercises totally different codepaths (goes straight to a LocalPantsRunner), which means that the second invocation will ignore the flag.
```

stuhood commented on src/python/pants/option/global_options.py:
```diff
@@ -222,6 +222,13 @@ def register_bootstrap_options(cls, register):
+    # This needs to be accessed at the same time as enable_pantsd,
+    # so we register it at bootstrap time.
+    register('--shutdown-pantsd-after-run', advanced=True, type=bool, default=False,
+      help='Shutdown pantsd after the current run.'
```
Need a space at the end of this line to avoid concatenating as `run.Will`.

The two sentences also don't seem super consistent... saying "will shutdown after this run" sortof implies that it is already running, but the next sentence says that that isn't legal. Would say something like: "starts, uses, and then stops pantsd for each run".

Also, rather than needing to clarify that this "won't work if pantsd was already running", it might be easier to just implement something whereby the client does something like:
```
if shutdown_pantsd_after_run and pantsd.running():
  pantsd.terminate()
```
I don't think that it would add much latency, assuming that the "running" check is just a PID lookup and check.

stuhood commented on src/python/pants/pantsd/pailgun_server.py:
```diff
@@ -99,7 +99,7 @@ class PailgunServer(ThreadingMixIn, TCPServer):
   daemon_threads = True
 
-  def __init__(self, server_address, runner_factory, lifecycle_lock,
+  def __init__(self, server_address, runner_factory, lifecycle_lock, request_complete_callback,
```
Worth expanding the docstring for this one I think.

blorente commented:
I have addressed the comments, and ended up overriding `ProcessManager::needs_restart` in `PantsDaemon`.
The only thing remaining is to make pantsd exit with _not an exception_ when one of the services dies, but because (1) it's behind an experimental flag, (2) I would need a bit of time to think about this and (3) this could be de-prioritized according to https://github.com/pantsbuild/pants/issues/7320#issuecomment-472211815, I'm inclined to leave that as a TODO.

stuhood commented:
Thanks!
